### ISSUE TITLE
GPU compositing bug on arm64

### ISSUE THREAD

**Issue by boazsender** [electron/electron#14246]
* Electron Version: [electron-v3.0.0-beta.5/electron-v3.0.0-beta.5-linux-armv7l](https://github.com/electron/electron/releases/download/v3.0.0-beta.5/electron-v3.0.0-beta.5-linux-armv7l.zip)
* Operating System (Platform and Version): [debian stretch for armhf](https://dl.vamrs.com/products/rock960/images/debian/rock960_debian_stretch_desktop_armhf_20180115.tar.gz) on a 64 bit arm single board computer
* Target hardware: [rock960](https://www.96boards.org/product/rock960/)
* Last known working Electron version: none

**Expected Behavior**
Expected electron to boot on it's default asar without crashing. Electron should be able to load https://get.webgl.org/ without crashing.

**Actual behavior**
When you boot electron on the target hardware it hangs and seems to be unable to do GPU compositing. When you disable GPU compositing, the system starts is able to start. Chromium 61 on the same board is able to GPU composite without issue.

**To Reproduce**
1. Get an arm 64 single board computer (we're using the rock 960)
2. Install the debian stretch for armhf (32 bit) image
3. open electron with it's default asar and watch it hang.

While we're using the rock960 (arm 64) board with the armhf build of debian stretch, we think this is reproducible on other arm64 boards running armhf. So far we've heard of this same bug reproducing with the same armhf build of debian stretch on a [Odroid XU4](https://www.hardkernel.com/main/products/prdt_info.php)

**Additional Information**
We realize that this may not be an electron bug, but rather an OS driver or other integration bug. We thought we should open this issue here in case this sounds familiar to anyone, or if there is anyone in the electron community who is able to help us further diagnose this issue.

Thank you for reading and for contributing to or maintaining the electron project! We really appreciate all your hard work, and are thankful for how great electron is for distributing node and chromium, and bringing the web platform to more places. Thanks again!

**Comment by Burnich** [comment]
Looks like the same problem as #12937

**Comment by jkleinsc** [comment]
@boazsender given that you are using an arm 64 computer have you tried using the arm64 build of Electron with arm64 debian stretch?

**Comment by boazsender** [comment]
@brunich
Thanks, it looks like #12937 is related. We tried out 1.7 and it is able to composite! That's a big relief for us, but we'd still like to be able to run 3.0. we are seeing some other GPU bugs, which we can potentially bhack around, but a 3.0/latest release seems much stabler.

@jkleinsc
We have not been able to find a arm64 distro with x. I think that might be related. We can keep looking, or poke around configuring our own windowing setup and test that, but getting armhf working would be great for us.

Thanks y'all for taking a look at this!

**Comment by boazsender** [comment]
Hi @sofianguy, thank you for triaging this issue. What further information do you need to unblock this issue?

**Comment by LemonAndroid** [comment]
Hey,

I'd love to tackle this bounty, but I want to confirm upfront, that no critical work has been done yet. So: Has someone already started work on this, put up a PR or even fixed the bug already?

On bountysource I saw that denimx2@bountysource.com started work ~1 month ago, but that doesn't have to mean anything.

**Comment by boazsender** [comment]
Awesome to hear @lemonandroid. As far as I know, no critical work. Has been done. I have not heard from @denimx2 since they started the bug.

Let me know how I can help.

**Comment by LemonAndroid** [comment]
Is emulation possible? I'm broke... :D

**Comment by LemonAndroid** [comment]
A quick google didn't yield any results for emulation.

I might be able to lend 100$ in case that is the correct item to acquire: https://store.vamrs.com/collections/rock960/products/rock960. Please confirm that the cheapest option is sufficient.

**Comment by boazsender** [comment]
You will likely be able to reproduce on raspberry pi, or another less expensive arm device running debian stretch armhf, but for this bug, we'd prefer to match the exact device just in case. If you need us to, we can ship you the device, though you ordering it directly is probably fastest. 

You can use the email in my profile.

Thank you!

**Comment by anszom** [comment]
Perhaps it isn't clear on BountySource, but the work here is almost done. I've made a fix and submitted a PR (see above) which is awaiting Electron developers' approval.

**Comment by modrobert** [comment]
> Perhaps it isn't clear on BountySource, but the work here is almost done. I've made a fix and submitted a PR (see above) which is awaiting Electron developers' approval.

Yes, apparently opened pull requests aren't shown at Bountysource, maybe they need to be closed on GitHub first?

I will look into this anyway, the rock960 is interesting, if nothing else for the learning process.

**Comment by iloseall** [comment]
I got the same problem.
Electron 3.0.10 and Electron 3.0.4
on raspberry pi 3B+ 
with  OS RASPBIAN STRETCH 
   OS Version:November 2018
   Kernel version:4.14

**Comment by modrobert** [comment]
If I start electron using this command it seems to run fine without patch:
`./electron 2>error.log`

**Comment by anszom** [comment]
Yes, this bug is manifested only when some of the standard file descriptors (0,1,2) point to a tty. Redirecting to a file or /dev/null is a verified workaround. A fix has already been commited to master, and is awaiting a merge to the 3.0 and 4.0 branch. I would assume that the fix will be included in the next release, but this depends on electron's release cycle, which I'm not familiar with.

**Comment by gnarf** [comment]
@nornagon is this released in a version I can test from npm yet?

**Comment by jkleinsc** [comment]
@gnarf it was released in [4.0.0-beta.8](https://github.com/electron/electron/releases/tag/v4.0.0-beta.8) and [3.0.11](https://github.com/electron/electron/releases/tag/v3.0.11).  So if you install electron@latest or electron@beta (or the specific versions above) the fix is there

**Comment by gnarf** [comment]
cheers thanks

**Comment by gev** [comment]
Just tested on version 3.1.8. Have a black screen after random time on worked application. Have same config too:

Raspberry pi 3B+
with OS RASPBIAN STRETCH
OS Version:November 2018
Kernel version:4.14

Will try ./electron 2>error.log

**Comment by gev** [comment]
I'd Add ./electron 2>error.lo. Also run all child processes with option { stdio: 'ignore' }. All works as well. Will to look for

**Comment by gev** [comment]
black screen is happenning still!

**Comment by gnarf** [comment]
The provided PR totally fixed the OP bug for us - so sorry @gev but I don't think your issue was a dupe of this one :(
